sintr_analysis_server_example/lib/instrumentation_query.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library sintr_worker_lib.instrumentation_query;
+
+import 'dart:async';
+
+import 'package:sintr_worker_lib/query.dart';
+import 'package:sintr_worker_lib/session_info.dart';
+
+final OPEN_BRACE = '{'.codeUnitAt(0);
+final QUOTE = '"'.codeUnitAt(0);
+
+/// Search the given [logMessageText] for the given [key]
+/// and return the associated value.
+/// This only returns simple values such as strings
+/// non-nested maps.
+String extractJsonValue(String logMessageText, String key) {
+  var prefix = '"$key"::';
+  int start = logMessageText.indexOf(prefix);
+  if (start == -1) throw 'expected $key in $logMessageText';
+  start += prefix.length;
+  int deliminator = logMessageText.codeUnitAt(start);
+  if (deliminator == QUOTE) {
+    // Return quoted string
+    ++start;
+    int end = logMessageText.indexOf('"', start);
+    if (end == -1) throw 'expected value for $key in $logMessageText';
+    return logMessageText.substring(start, end);
+  } else if (deliminator == OPEN_BRACE) {
+    // Return JSON map
+    // TODO (danrubel) handle nested and embedded braces
+    int end = logMessageText.indexOf('}');
+    if (end == -1) {
+      throw 'expected matching brace for $key in $logMessageText';
+    }
+    return logMessageText.substring(start, end + 1);
+  }
+  // TODO (danrubel) handled ints and lists
+  return null;
+}
+
+/// Insert [newValue] into the sorted list of [values]
+/// such that the list is still sorted.
+void orderedInsert(List values, var newValue, [int comparator(v1, v2)]) {
+  if (values.length == 0) {
+    values.add(newValue);
+    return;
+  }
+  if (comparator == null) comparator = (v1, v2) => v1 - v2;
+  if (comparator(newValue, values[0]) < 0) {
+    values.insert(0, newValue);
+    return;
+  }
+  int start = 0;
+  int end = values.length;
+  int pivot = start + (end - start) ~/ 2;
+  while (end - start > 1 || values[pivot] == newValue) {
+    var diff = comparator(newValue, values[pivot]);
+    if (diff == 0) break;
+    if (diff > 0) {
+      start = pivot;
+    } else {
+      end = pivot;
+    }
+    pivot = start + (end - start) ~/ 2;
+  }
+  values.insert(pivot + 1, newValue);
```

Comment:
Consider adding a post completion invariant check behind a DEBUG flag that the list is sorted.